servers/plugins/org.fusesource.ide.server.karaf.core/src/org/fusesource/ide/server/karaf/core/publish/jmx/KarafJMXPublisher.java

```diff
@@ -200,11 +209,49 @@ private boolean reinstallBundle(IServer server, long bundleId, IPath file) throw
 	 * @param module
 	 * @return
 	 */
-	private long installBundle(IServer server, IPath file) throws CoreException {
+	private long installBundle(IServer server, IPath file) {
 		if (file != null) {
-			long bundleId = this.jmxPublisher.installBundle(mbsc, file.toOSString().replace("file:", ""));
-			return bundleId;
+			return this.jmxPublisher.installBundle(mbsc, file.toOSString().replace("file:", ""));
 		}
 		return -1;
 	}
+	
+	protected void refreshJMXView() {
+		Display.getDefault().asyncExec( () -> {
+			CommonNavigator nav = getNavigator();
+			Root root = getAffectedServerEntry(nav);
+			if (root != null) {
+				root.refresh();
+			}
+			if (nav != null) {
+				nav.getCommonViewer().refresh();
+			}
+		});
+	}
+	
+	private Root getAffectedServerEntry(CommonNavigator nav) {
+		if (nav != null) {
+			nav.getCommonViewer().refresh();
+			for (Object o : nav.getCommonViewer().getExpandedElements()) {
+				if (o instanceof IConnectionWrapper) {
+					IConnectionWrapper cw = (IConnectionWrapper)o;
+					return cw.getRoot();
+				}
+			}
+		}
+		return null;
+	}
+	
+	private CommonNavigator getNavigator() {
+		if (PlatformUI.getWorkbench().getActiveWorkbenchWindow() != null) {
+			IWorkbenchPage page = PlatformUI.getWorkbench().getActiveWorkbenchWindow().getActivePage();
+			if (page != null) {
+				IViewPart viewPart = page.findView(UIHelper.ID_SERVERS_VIEW);
```

Comment:
Because you are usually using Add/Remove Software from inside the servers view. Besides that the IConnectionWrapper instance is shared between both views so I don't see a difference.

Reply:
in theory, we should be able to Run as -> Run on server https://tools.jboss.org/documentation/howto/servers_deploytolocalserver.html#Deploy
but it seems that it wasn't implemented.
The day we will implement it to be fully aligned with JBoss Tools, by implementing it relying on the Servers view it won't work.

If I modify a project, it is automatically redeployed, if the Servers view is closed and JMX view opened, the refresh does not occur.

Reply:
If you open the JMX node on the JMX Navigator then it is also connected in the servers view. We are just sharing navigator content here.

Reply:
How can it be connected to Servers View if Servers view is not opened? it creates a ghost instance?

Reply:
This PR does a JMX tree refresh (on a particular instance of IConnectionWrapper which represents the connected server) by issuing a reload for this connection wrapper instance whenever someone is adding or removing a module in the servers view. There is no other way than maybe autodeployment when changing sources doing this. I see your point but I don't see any better solution yet.